### ISSUE TITLE
Interactive Machine Learning Experiments | Trekhleb

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7814]
Interactive Machine Learning Experiments | Trekhleb<br>
<br>
https://ift.tt/3rZf8zh<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<article>
<h2>TL;DR</h2>
<p>Hey readers!</p>
<p>I've open-sourced new <a href="https://github.com/trekhleb/machine-learning-experiments">&#129302; Interactive Machine Learning Experiments</a> project on GitHub. Each experiment consists of &#127947;&#65039; <em>Jupyter/Colab notebook</em> (to see how a model was trained) and &#127912; <em>demo page</em> (to see a model in action right in your browser).</p>
<p>Although the models may be a little dumb (remember, these are just experiments, not a production ready code), they will try to do their best to:</p>
<ul>
<li>&#128396; Recognize digits or sketches you draw in your browser</li>
<li>&#128248; Detect and recognize the objects you'll show to your camera</li>
<li>&#127749; Classify your uploaded image</li>
<li>&#128221; Write a Shakespeare poem with you</li>
<li>&#9994;&#128400;&#9996;&#65039; Play with you in Rock-Paper-Scissors game</li>
<li>etc.</li>
</ul>
<p>I've trained the models on <em>Python</em> using <em>TensorFlow 2</em> with <em>Keras</em> support and then consumed them for a demo in a browser using <em>React</em> and <em>JavaScript</em> version of <em>Tensorflow</em>.</p>
<p><span><img alt="Interactive Machine Learning Experiments" title="Interactive Machine Learning Experiments" src="https://trekhleb.dev/static/3e4b832a607b38fe6ebf64a74304925e/00d43/13.png"></span></p>
<h2>Models performance</h2>
<p>&#9888;&#65039; First, let's set our expectations.&#65039; The repository contains machine learning <strong>experiments</strong> and <strong>not</strong> a production ready, reusable, optimised and fine-tuned code and models. This is rather a sandbox or a playground for learning and trying different machine learning approaches, algorithms and data-sets. Models might not perform well and there is a place for overfitting/underfitting.</p>
<p>Therefore, sometimes you might see things like this:</p>
<p><span><img alt="Dumb model" title="Dumb model" src="https://trekhleb.dev/static/07973cc81709f3ceb90984bf76ea68b7/00d43/0.png"></span></p>
<p>But be patient, sometimes the model might get smarter &#129299; and give you this:</p>
<p><span><img alt="Smart model" title="Smart model" src="https://trekhleb.dev/static/f0c246e87d8461bedaac585f7531b9b0/00d43/1.png"></span></p>
<h2>Background</h2>
<p>I'm a <a href="https://www.linkedin.com/in/trekhleb/">software engineer</a> and for the last several years now I've been doing mostly frontend and backend programming. In my spare time, as a hobby, I decided to dig into machine learning topics to make it less <em>like magic</em> and <em>more like math</em> to myself.</p>
<ol>
<li>
<p>&#128467; Since <strong>Python</strong> might be a good choice to start experimenting with Machine Learning I decided to learn its basic syntax first. As a result a <a href="https://github.com/trekhleb/learn-python">&#128013; Playground and Cheatsheet for Learning Python</a> project came out. This was just to practice Python and at the same time to have a cheatsheet of basic syntax once I need it (for things like <code>dict_via_comprehension = {x: x**2 for x in (2, 4, 6)}</code> etc.).</p>
</li>
<li>
<p>&#128467; After learning a bit of Python I wanted to dig into the basic <strong>math</strong> behind Machine Learning. So after passing an awesome <a href="https://www.coursera.org/learn/machine-learning">Machine Learning course by Andrew Ng</a> on Coursera the <a href="https://github.com/trekhleb/homemade-machine-learning">&#129302; Homemade Machine Learning</a> project came out. This time it was about creating a cheatsheet for basic machine learning math algorithms like linear regression, logistic regression, k-means, multilayer perceptron etc.</p>
</li>
<li>
<p>&#128467; The next attempt to play around with basic Machine Learning math was <a href="https://github.com/trekhleb/nano-neuron">&#129302; NanoNeuron</a>. It was about 7 simple JavaScript functions that supposed to give you a feeling of how machines can actually "learn".</p>
</li>
<li>
<p>&#128467; After finishing yet another awesome <a href="https://www.coursera.org/specializations/deep-learning">Deep Learning Specialization by Andrew Ng</a> on Coursera I decided to practice a bit more with <strong>multilayer perceptrons</strong>, <strong>convolutional</strong> and <strong>recurrent neural networks</strong> (CNNs and RNNs). This time instead of implementing everything from scratch I decided to start using some machine learning framework. I ended up using <a href="https://www.tensorflow.org/">TensorFlow 2</a> with <a href="https://www.tensorflow.org/guide/keras/overview">Keras</a>. I also didn't want to focus too much on math (letting the framework do it for me) and instead I wanted to come up with something more practical, applicable and something I could try to play with right in my browser. As a result new <a href="https://github.com/trekhleb/machine-learning-experiments">&#129302; Interactive Machine Learning Experiments</a> came out that I want to describe a bit more here.</p>
</li>
</ol>
<h2>Tech-stack</h2>
<h3>Models training</h3>
<ul>
<li>
<p>&#127947;&#127995;&zwj; I used <a href="https://www.tensorflow.org/guide/keras/overview">Keras</a> inside <a href="https://www.tensorflow.org/">TensorFlow 2</a> for modelling and training. Since I had zero experience with machine learning frameworks, I needed to start with something. One of the selling points in favor of TensorFlow was that it has both Python and <a href="https://www.tensorflow.org/js">JavaScript flavor</a> of the library with similar API. So eventually I used Python version for training and JavaScript version for demos.</p>
</li>
<li>
<p>&#127947;&#127995;&zwj; I trained TensorFlow models on Python inside <a href="https://jupyter.org/">Jupyter</a> notebooks locally and sometimes used <a href="https://colab.research.google.com/">Colab</a> to make the training faster on GPU.</p>
</li>
<li>
<p>&#128187; Most of the models were trained on good old MacBook's Pro CPU (2,9 GHz Dual-Core Intel Core i5).</p>
</li>
<li>
<p>&#128290; Of course there is no way you could run away from <a href="https://numpy.org/">NumPy</a> for matrix/tensors operations.</p>
</li>
</ul>
<h3>Models demo</h3>
<ul>
<li>
<p>&#127947;&#127995;&zwj; I used <a href="https://www.tensorflow.org/js">TensorFlow.js</a> to do predictions with previously trained models.</p>
</li>
<li>
<p>&#9851;&#65039; To convert <em>Keras HDF5</em> models to <em>TensorFlow.js Layers</em> format I used <a href="https://github.com/tensorflow/tfjs/tree/master/tfjs-converter">TensorFlow.js converter</a>. This might be inefficient to transfer the whole model (megabytes of data) to the browser instead of making predictions through HTTP requests, but again, remember that these are just experiments and not production-ready code and architecture. I wanted to avoid having a dedicated back-end service to make architecture simpler.</p>
</li>
<li>
<p>&#128104;&#127995;&zwj;&#127912; The <a href="http://trekhleb.github.io/machine-learning-experiments">Demo application</a> was created on <a href="https://reactjs.org/">React</a> using <a href="https://github.com/facebook/create-react-app">create-react-app</a> starter with a default <a href="https://flow.org/en/">Flow</a> flavour for type checking.</p>
</li>
<li>
<p>&#128133;&#127995; For styling, I used <a href="https://material-ui.com/">Material UI</a>. It was, as they say, "to kill two birds" at once and try out a new styling framework (sorry, <a href="https://getbootstrap.com/">Bootstrap</a> &#129335;&#127995;&zwj;).</p>
</li>
</ul>
<h2>Experiments</h2>
<p>So, in short, you may access Demo page and Jupyter notebooks by these links:</p>
<h3>Experiments with Multilayer Perceptron (MLP)</h3>
<blockquote>
<p>A <a href="https://en.wikipedia.org/wiki/Multilayer_perceptron">multilayer perceptron (MLP)</a> is a class of feedforward artificial neural network (ANN). Multilayer perceptrons are sometimes referred to as "vanilla" neural networks (composed of multiple layers of perceptrons), especially when they have a single hidden layer.</p>
</blockquote>
<h4>Handwritten Digits Recognition</h4>
<p>You draw a digit, and the model tries to recognize it.</p>
<p><img alt="Handwritten Digits Recognition" src="https://trekhleb.dev/posts-assets/ddeabb820e5eaf3a6737ea3798805b22/2.gif"></p>
<h4>Handwritten Sketch Recognition</h4>
<p>You draw a sketch, and the model tries to recognize it.</p>
<p><img alt="Handwritten Sketch Recognition" src="https://trekhleb.dev/posts-assets/a1af98f24f9687516f436c3997306916/3.gif"></p>
<h3>Experiments with Convolutional Neural Networks (CNN)</h3>
<blockquote>
<p>A <a href="https://en.wikipedia.org/wiki/Convolutional_neural_network">convolutional neural network (CNN, or ConvNet)</a> is a class of deep neural networks, most commonly applied to analyzing visual imagery (photos, videos). They are used for detecting and classifying objects on photos and videos, style transfer, face recognition, pose estimation etc.</p>
</blockquote>
<h4>Handwritten Digits Recognition (CNN)</h4>
<p>You draw a digit, and the model tries to recognize it. This experiment is similar to the one from MLP section, but it uses CNN under the hood.</p>
<p><img alt="Handwritten Digits Recognition (CNN)" src="https://trekhleb.dev/posts-assets/ddeabb820e5eaf3a6737ea3798805b22/4.gif"></p>
<h4>Handwritten Sketch Recognition (CNN)</h4>
<p>You draw a sketch, and the model tries to recognize it. This experiment is similar to the one from MLP section, but it uses CNN under the hood.</p>
<p><img alt="Handwritten Sketch Recognition (CNN)" src="https://trekhleb.dev/posts-assets/a1af98f24f9687516f436c3997306916/5.gif"></p>
<h4>Rock Paper Scissors (CNN)</h4>
<p>You play a Rock-Paper-Scissors game with the model. This experiment uses CNN that is trained from scratch.</p>
<p><img alt="Rock Paper Scissors (CNN)" src="https://trekhleb.dev/posts-assets/ba1ce2c409ef1f859f2379aa913b884a/6.gif"></p>
<h4>Rock Paper Scissors (MobilenetV2)</h4>
<p>You play a Rock-Paper-Scissors game with the model. This model uses transfer learning and is based on <a href="https://www.tensorflow.org/api_docs/python/tf/keras/applications/MobileNetV2">MobilenetV2</a>.</p>
<p><img alt="Rock Paper Scissors (MobilenetV2)" src="https://trekhleb.dev/posts-assets/62bd44fd2baf2044c85dcf300ca340e2/7.gif"></p>
<h4>Objects Detection (MobileNetV2)</h4>
<p>You show to the model your environment through your camera, and it will try to detect and recognize the objects. This model uses transfer learning and is based on <a href="https://www.tensorflow.org/api_docs/python/tf/keras/applications/MobileNetV2">MobilenetV2</a>.</p>
<p><img alt="Objects Detection (MobileNetV2)" src="https://trekhleb.dev/posts-assets/85eceefa39aff6781db023fe14c41d3e/8.gif"></p>
<h4>Image Classification (MobileNetV2)</h4>
<p>You upload a picture, and the model tries to classify it depending on what it "sees" on the picture. This model uses transfer learning and is based on <a href="https://www.tensorflow.org/api_docs/python/tf/keras/applications/MobileNetV2">MobilenetV2</a>.</p>
<p><img alt="Image Classification (MobileNetV2)" src="https://trekhleb.dev/posts-assets/687de96293f15861031ad7fc1b6da1ed/9.gif"></p>
<h3>Experiments with Recurrent Neural Networks (RNN)</h3>
<blockquote>
<p>A <a href="https://en.wikipedia.org/wiki/Recurrent_neural_network">recurrent neural network (RNN)</a> is a class of deep neural networks, most commonly applied to sequence-based data like speech, voice, text or music. They are used for machine translation, speech recognition, voice synthesis etc.</p>
</blockquote>
<h4>Numbers Summation</h4>
<p>You type a summation expression (i.e. <code>17+38</code>), and the model predicts the result (i.e. <code>55</code>). The interesting part here is that the model treats the input as a <em>sequence</em>, meaning it learned that when you type a sequence <code>1</code> &rarr; <code>17</code> &rarr; <code>17+</code> &rarr; <code>17+3</code> &rarr; <code>17+38</code> it "translates" it to another sequence <code>55</code>. You may think about it as translating a Spanish <code>Hola</code> sequence to English <code>Hello</code>.</p>
<p><img alt="Numbers Summation" src="https://trekhleb.dev/posts-assets/e493ba110b3512b3c8367d0c1287905f/10.gif"></p>
<h4>Shakespeare Text Generation</h4>
<p>You start typing a poem like Shakespeare, and the model will continue it like Shakespeare. At least it will try to do so &#128512;.</p>
<p><img alt="Shakespeare Text Generation" src="https://trekhleb.dev/posts-assets/93ebaea4c757d34a8d9ba02e705fcaa9/11.gif"></p>
<h4>Wikipedia Text Generation</h4>
<p>You start typing a Wiki article, and the model tries to continue it.</p>
<p><img alt="Wikipedia Text Generation" src="https://trekhleb.dev/posts-assets/9efc3106b06044a6e4f6ff91e2d5cbbc/12.gif"></p>
<h2>Future plans</h2>
<p>As I've mentioned above the main purpose of <a href="https://github.com/trekhleb/machine-learning-experiments">the repository</a> is to be more like a playground for learning rather than for production-ready models. Therefore, the main plan is to <strong>continue learning and experimenting</strong> with deep-learning challenges and approaches. The next interesting challenges to play with might be:</p>
<ul>
<li>Emotions detection</li>
<li>Style transfer</li>
<li>Language translation</li>
<li>Generating images (i.e. handwritten numbers)</li>
<li>etc.</li>
</ul>
<p>Another interesting opportunity would be to <strong>tune existing models to make them more performant</strong>. I believe it might give a better understanding of how to overcome overfitting and underfitting and what to do with the model if it just stuck on <code>60%</code> accuracy level for both training and validation sets and doesn't want to improve anymore &#129300;.</p>
<p>Anyways, I hope you might find some useful insights for models training from <a href="https://github.com/trekhleb/machine-learning-experiments">the repository</a> or at least to have some fun playing around with the demos!</p>
<p>Happy learning! &#129302;</p>
</article>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Trekhleb<br>
<br>
April 25, 2021 at 03:08PM
